lib/chibi/json.c

```diff
@@ -307,10 +317,267 @@ sexp sexp_parse_json (sexp ctx, sexp self, sexp_sint_t n, sexp str) {
+        res = sexp_json_unparse_exception(ctx, self, "unable to encode string", obj);
+        sexp_gc_release2(ctx);
+        return res;
+      }
+      sprintf(cout, "\\u%04lX\\u%04lX", chh, chl);
+    }
+    tmp = sexp_c_string(ctx, cout, -1);
+    res = sexp_cons(ctx, tmp, res);
+  }
+
+  tmp = sexp_c_string(ctx, "\"", -1);
+  res = sexp_cons(ctx, tmp, res);
+
+  res = sexp_nreverse(ctx, res);
+  res = sexp_string_concatenate(ctx, res, SEXP_FALSE);
+
+  sexp_gc_release2(ctx);
+  return res;
+}
+
+sexp unparse_json_array(sexp ctx, sexp self, const sexp obj) {
+  sexp_gc_var2(res, tmp);
+  sexp_gc_preserve2(ctx, res, tmp);
+  res = SEXP_NULL;
+
+  tmp = sexp_c_string(ctx, "[", -1);
+  res = sexp_cons(ctx, tmp, res);
+
+  int len = sexp_vector_length(obj);
+  for (int i=0; i!=len; i++) {
+    tmp = unparse_json(ctx, self, sexp_vector_ref(obj, sexp_make_fixnum(i)));
+    if (sexp_exceptionp(tmp)) {
+      sexp_gc_release2(ctx);
+      return tmp;
+    }
+    res = sexp_cons(ctx, tmp, res);
+
+    if (i != len-1) {
+      tmp = sexp_c_string(ctx, ",", -1);
+      res = sexp_cons(ctx, tmp, res);
+    }
+  }
+
+  tmp = sexp_c_string(ctx, "]", -1);
+  res = sexp_cons(ctx, tmp, res);
+
+  res = sexp_nreverse(ctx, res);
+  res = sexp_string_concatenate(ctx, res, SEXP_FALSE);
+
+  sexp_gc_release2(ctx);
+  return res;
+}
+
+
+sexp unparse_json_object(sexp ctx, sexp self, const sexp obj) {
+  sexp_gc_var6(res, tmp, it, cur, key, val);
+  sexp_gc_preserve6(ctx, res, tmp, it, cur, key, val);
+  res = SEXP_NULL;
+
+  tmp = sexp_c_string(ctx, "{", -1);
+  res = sexp_cons(ctx, tmp, res);
+
+  int len = sexp_unbox_fixnum(sexp_length(ctx, obj));
+  it = obj;
+  for (int i=0; i!=len; i++) {
+    cur = sexp_car(it);
+    if (!sexp_pairp(cur)) {
+      res = sexp_json_unparse_exception(ctx, self, "unable to encode key-value pair: not a pair", obj);
+      goto except;
+    }
+
+    // Key
+    key = sexp_car(cur);
+    if (!(sexp_symbolp(key) /*|| sexp_stringp(key)*/)) {
+      res = sexp_json_unparse_exception(ctx, self, "unable to encode key: not a symbol", key);
+      goto except;
+    }
+    tmp = unparse_json(ctx, self, key);
+    if (sexp_exceptionp(tmp)) {
+      res = tmp;
+      goto except;
+    }
+    res = sexp_cons(ctx, tmp, res);
+
+    // Separator
+    tmp = sexp_c_string(ctx, ":", -1);
+    res = sexp_cons(ctx, tmp, res);
+
+    // Value
+    val = sexp_cdr(cur);
+    tmp = unparse_json(ctx, self, val);
+    res = sexp_cons(ctx, tmp, res);
+
+    if (i != len-1) {
+      tmp = sexp_c_string(ctx, ",", -1);
+      res = sexp_cons(ctx, tmp, res);
+    }
+    it = sexp_cdr(it);
+  }
+
+  tmp = sexp_c_string(ctx, "}", -1);
+  res = sexp_cons(ctx, tmp, res);
+
+  res = sexp_nreverse(ctx, res);
+  res = sexp_string_concatenate(ctx, res, SEXP_FALSE);
+except:
+  sexp_gc_release6(ctx);
+  return res;
+}
+
+sexp unparse_json (sexp ctx, sexp self, sexp obj) {
+  sexp_gc_var1(res);
+  sexp_gc_preserve1(ctx, res);
+  res = SEXP_NULL;
```

Comment:
The default return value should be SEXP_VOID.

Reply:
Done in 28f3641 
Thanks!